### PR TITLE
feat: invoke `handle_batch_open_requests`

### LINKED ISSUE #4026: [Tracking issue] open regions in batches
### What type of enhancement is this?

Performance, Refactor, Tech debt reduction

### What does the enhancement do?
- [x] #4027 
- [x] Refactor `LogStore`, Add the `WalEntryReader` to decode raw WAL entries into upper-level WAL entries #4038
- [x] Add the `RawEntryReader` to read undecoded raw WAL entries #4030
- [x] #4031
- [x] Open a region in the background #4052 
- [x] Add the `handle_batch_open_requests` to the `RegionEngine` trait and implement the `handle_batch_open_requests` method for the mito engine #4075
- [ ] Add the `handle_batch_open_requests` to the region server  #4107

The whole picture:

image

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close #4026 

## What's changed and what's your intention?

Via invoking `handle_batch_open_requests,` which allows the database to read the WAL entries once, to open all regions sharing the same WAL topic.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/datanode/src/datanode.rs [hunk 1]
```diff
@@ -468,39 +467,46 @@ async fn open_all_regions(
...
+                region_dir,
+                options,
+                skip_wal_replay: false,
+            },
+        ));
     }
-    let _ = try_join_all(tasks).await?;
 
+    let open_regions = region_server
+        .handle_batch_open_requests(OPEN_REGION_PARALLELISM, region_requests)
```
Comment (fengjiachun) [inline src/datanode/src/datanode.rs:488]: Can we use startup parameter to set the parallelism?

File: src/datanode/src/datanode.rs [hunk 2]
```diff
@@ -468,39 +467,46 @@ async fn open_all_regions(
         }
     }
     info!("going to open {} region(s)", regions.len());
-    let semaphore = Arc::new(tokio::sync::Semaphore::new(OPEN_REGION_PARALLELISM));
-    let mut tasks = vec![];
 
-    let region_server_ref = &region_server;
+    let mut region_requests = Vec::with_capacity(regions.len());
+    let mut region_ids = Vec::with_capacity(regions.len());
```
Comment (fengjiachun) [inline src/datanode/src/datanode.rs:472]: It seems unnecessary, the `handle_batch_open_requests` method will return region ids.

**Review by fengjiachun — APPROVED** [review]
Will definitely be LGTM

File: src/datanode/src/region_server.rs [hunk 3]
```diff
@@ -454,6 +467,97 @@ impl RegionServerInner {
...
+                    match result {
+                        Ok(_) => {
+                            let _ = self
+                                .set_region_status_ready(region_id, engine.clone(), *region_change)
+                                .await;
+                            open_regions.push(region_id)
+                        }
+                        Err(err) => {
+                            self.unset_region_status(region_id, *region_change);
+                            error!("Failed to open region: {}, error: {err}", region_id);
```
Comment (evenyag) [inline src/datanode/src/region_server.rs:506]:
```suggestion
                            error!(e; "Failed to open region: {}", region_id);
```

File: src/datanode/src/region_server.rs [hunk 4]
```diff
@@ -454,6 +467,97 @@ impl RegionServerInner {
...
+            .with_context(|_| HandleBatchOpenRequestSnafu)
+        {
+            Ok(results) => {
+                for (region_id, result) in results {
+                    let region_change = &region_changes[&region_id];
+                    match result {
+                        Ok(_) => {
+                            let _ = self
+                                .set_region_status_ready(region_id, engine.clone(), *region_change)
+                                .await;
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/datanode/src/region_server.rs]: Now, it throws errors

**Pull request merged**